File: crawler/geocoding.py
"""
Geocoding Module - 주소를 위도/경도로 변환
"""
import requests
import time
import logging
from config import KAKAO_REST_API_KEY, REQUEST_TIMEOUT

logger = logging.getLogger(__name__)


def geocode_address(address: str) -> dict:
    """
    Kakao Geocoding API를 사용하여 주소를 위도/경도로 변환

    Args:
        address: 주소 문자열

    Returns:
        {'lat': float, 'lng': float} 또는 None
    """
    if not address or not address.strip():
        logger.warning("Empty address provided")
        return None

    if not KAKAO_REST_API_KEY:
        logger.error("Kakao REST API Key not configured")
        return None

    url = "https://dapi.kakao.com/v2/local/search/address.json"
    headers = {"Authorization": f"KakaoAK {KAKAO_REST_API_KEY}"}
    params = {"query": address.strip()}

    try:
        response = requests.get(
            url,
            headers=headers,
            params=params,
            timeout=REQUEST_TIMEOUT
        )
        response.raise_for_status()
        data = response.json()

        if data.get('documents'):
            doc = data['documents'][0]
            result = {
                'lat': float(doc['y']),
                'lng': float(doc['x'])
            }
            logger.info(f"Geocoded: {address} -> ({result['lat']}, {result['lng']})")
            return result
        else:
            logger.warning(f"No geocoding result for address: {address}")
            return None

    except requests.exceptions.Timeout:
        logger.error(f"Geocoding timeout for address: {address}")
        return None
    except requests.exceptions.RequestException as e:
        logger.error(f"Geocoding request failed: {e}")
        return None
    except (KeyError, ValueError, IndexError) as e:
        logger.error(f"Geocoding response parsing error: {e}")
        return None


def geocode_batch(addresses: list, delay: float = 0.1) -> dict:
    """
    여러 주소를 배치로 Geocoding

    Args:
        addresses: 주소 리스트
        delay: API 호출 간 딜레이 (초)

    Returns:
        {address: {lat, lng}} 딕셔너리
    """
    results = {}

    for i, address in enumerate(addresses):
        if address in results:
            continue

        logger.info(f"Geocoding {i+1}/{len(addresses)}: {address}")
        result = geocode_address(address)

        if result:
            results[address] = result
        else:
            results[address] = None

        # Rate limiting
        if i < len(addresses) - 1:
            time.sleep(delay)

    success_count = sum(1 for v in results.values() if v is not None)
    logger.info(f"Geocoding complete: {success_count}/{len(addresses)} successful")

    return results

File: src/lib/geocoding.ts
/**
 * Kakao Geocoding API 유틸리티
 * 주소를 좌표(위도, 경도)로 변환합니다.
 */

type GeocodingResult = {
  latitude: number;
  longitude: number;
  address: string;
};

type KakaoGeocodingResponse = {
  meta: {
    total_count: number;
  };
  documents: Array<{
    address_name: string;
    x: string; // 경도
    y: string; // 위도
  }>;
};

/**
 * 주소를 좌표로 변환하는 함수
 * @param address 변환할 주소
 * @returns 좌표 정보 (위도, 경도)
 */
export async function geocodeAddress(
  address: string
): Promise<GeocodingResult | null> {
  const apiKey = process.env.KAKAO_REST_API_KEY;

  if (!apiKey) {
    console.error('KAKAO_REST_API_KEY가 설정되지 않았습니다.');
    return null;
  }

  try {
    const encodedAddress = encodeURIComponent(address);
    const url = `https://dapi.kakao.com/v2/local/search/address.json?query=${encodedAddress}`;

    const response = await fetch(url, {
      headers: {
        Authorization: `KakaoAK ${apiKey}`,
      },
    });

    if (!response.ok) {
      console.error(
        `Geocoding API 오류: ${response.status} ${response.statusText}`
      );
      return null;
    }

    const data: KakaoGeocodingResponse = await response.json();

    if (data.meta.total_count === 0) {
      console.warn(`좌표를 찾을 수 없는 주소: ${address}`);
      return null;
    }

    const document = data.documents[0];

    return {
      latitude: parseFloat(document.y),
      longitude: parseFloat(document.x),
      address: document.address_name,
    };
  } catch (error) {
    console.error('Geocoding 오류:', error);
    return null;
  }
}

/**
 * 여러 주소를 배치로 변환하는 함수 (Rate Limiting 고려)
 * @param addresses 주소 배열
 * @param delayMs 각 요청 간 지연 시간 (기본 100ms)
 * @returns 좌표 정보 배열
 */
export async function geocodeAddressBatch(
  addresses: string[],
  delayMs: number = 100
): Promise<(GeocodingResult | null)[]> {
  const results: (GeocodingResult | null)[] = [];

  for (const address of addresses) {
    const result = await geocodeAddress(address);
    results.push(result);

    // Rate Limiting 방지를 위한 지연
    if (addresses.indexOf(address) < addresses.length - 1) {
      await new Promise((resolve) => setTimeout(resolve, delayMs));
    }
  }

  return results;
}
